### DISCUSSION TITLE
Add new API to query user data

### DISCUSSION THREAD

**Discussion by Aleson95** [keycloak/keycloak#14272] [Q&A] [answered]
Hello friend,

What SPI should I use when I wanted to add new API to query the user data?
Is it RealmResourceProvider or User Storage SPI ?

**Comment by dishi-github** [comment, marked as answer]
User Storage SPI if you are using Keycloak v 18.0 or below. For higher versions, Map Storage SPI is under development. User Storage SPI  can still be used in higher versions with legacy binaries.

**Reply by Aleson95** [reply]
I see thank you friend for the reply
